pkg/hubble/exporter/dynamic_exporter.go

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package exporter
+
+import (
+	"context"
+	"errors"
+
+	"github.com/sirupsen/logrus"
+
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/hubble/exporter/exporteroption"
+	"github.com/cilium/cilium/pkg/hubble/metrics"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const gaugeMetricReportInterval = 5 * time.Second
+
+type dynamicExporter struct {
+	logger           logrus.FieldLogger
+	watcher          *configWatcher
+	managedExporters map[string]*managedExporter
+	maxFileSizeMB    int
+	maxBackups       int
+	mutex            lock.RWMutex
+	metricsTicker    *time.Ticker
+	metricsStop      chan bool
+}
+
+// OnDecodedEvent distributes events across all managed exporters.
+func (d *dynamicExporter) OnDecodedEvent(ctx context.Context, event *v1.Event) (bool, error) {
+	select {
+	case <-ctx.Done():
+		return false, d.Stop()
+	default:
+	}
+
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	for _, me := range d.managedExporters {
+		if me.config.End.After(time.Now()) {
+			me.exporter.OnDecodedEvent(ctx, event)
+		}
+	}
+	return false, nil
+}
+
+// Stop stops configuration watcher, metrics propagation and all managed flow
+// log exporters.
+func (d *dynamicExporter) Stop() error {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	d.watcher.Stop()
+	d.metricsTicker.Stop()
+	d.metricsStop <- true
+
+	var accErrs []error
+	for _, me := range d.managedExporters {
+		accErrs = append(accErrs, me.exporter.Stop())
+	}
+
+	return errors.Join(accErrs...)
+}
+
+// NewDynamicExporter creates instance of dynamic hubble flow exporter.
+func NewDynamicExporter(configFilePath string, maxFileSizeMB, maxBackups int) *dynamicExporter {
+	dynamicExporter := &dynamicExporter{
+		logger:           logrus.New(),
+		managedExporters: make(map[string]*managedExporter),
+		maxFileSizeMB:    maxFileSizeMB,
+		maxBackups:       maxBackups,
+	}
+
+	watcher := NewConfigWatcher(configFilePath, dynamicExporter.onConfigReload)
+	dynamicExporter.watcher = watcher
+	dynamicExporter.startMetricReporting()
+	return dynamicExporter
+}
+
+func (d *dynamicExporter) startMetricReporting() {
+	d.metricsTicker = time.NewTicker(gaugeMetricReportInterval)
+	d.metricsStop = make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-d.metricsStop:
+				return
+			case <-d.metricsTicker.C:
+				d.reportGaugeMetrics()
+			}
+		}
+	}()
+}
+
+func (d *dynamicExporter) reportGaugeMetrics() {
```

Comment:
I'm not a huge fan of metrics being updated asynchronously. Can you update the metrics when the exports are updated instead?

Reply:
Some metrics can be updated after reconfiguration. I added a call to update metrics after reconfiguration. For active/inactive metrics they depend on current time and declared flowlog end date so it needs to be recalculated asynchronously even when no reconfiguration is applied.

Reply:
Ah, I see what you mean. In that case, it might make more sense to implement the `Collector` interface for metrics and report the state at collection time, rather than doing it periodically. In this model, `Collect()` is called and you send the 'current' values on the provided channel.

Reply:
Fixed